### PR TITLE
feat(RevenueCatUI): add presentation-session state store (PWENG-57)

### PR DESCRIPTION
**Pull request by AlvaroBrey**
Introduces `PaywallStateStore`, the runtime state container. Seeded from `state_declarations` defaults; `applyUpdates` writes to it and `currentValueOrDefault` reads with default fallback.

Scoping: standalone paywalls own one store for the duration of the presentation; workflow paywalls share one store across all screens. Re-registering a screen's declarations (on navigation) preserves any values already written by other screens.

Internally uses `mutableStateMapOf<String, MutableState<JsonPrimitive>>` so a write to one key only invalidates `derivedStateOf` blocks reading that key, not every component. `declaredTypes` is a plain `ConcurrentHashMap` (never snapshot-read).

---

> [!NOTE]
> **Medium Risk**
> New shared mutable state across workflow steps and async prewarm adds concurrency/session-boundary edge cases; changes are localized to RevenueCatUI paywall presentation with tests, but incorrect scoping could leak or drop user-driven state between screens.
> 
> **Overview**
> Adds **`PaywallStateStore`**, a per-presentation-session container for dashboard **state-driven** paywall values: seeded from `state_declarations`, updated via typed `StateUpdate` writes (including payload references), and read through `currentValueOrDefault`. Compose uses per-key `MutableState` so updates only invalidate readers of that key.
> 
> **`PaywallState.Loaded.Components`** now carries a `stateStore`. Validation/mapping threads **`stateDeclarations`** from paywall components data into component paywall state creation.
> 
> **`PaywallViewModel`** owns session scoping: one reused store for standalone paywalls across refreshes (e.g. color scheme) until dismiss; one shared store per workflow impression across steps/screens, with declarations registered as each step builds. Prewarm captures the workflow store and bails if a newer session replaces it, avoiding stale cache pollution.
> 
> Unit tests cover store behavior and ViewModel reuse/session-swap edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e436d71eaffdf7c7a629fc9f8d17a22f83974cb5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by AlvaroBrey** [comment]
* **#3664** <a href="https://app.graphite.com/github/pr/RevenueCat/purchases-android/3664?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3661** <a href="https://app.graphite.com/github/pr/RevenueCat/purchases-android/3661?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3660** <a href="https://app.graphite.com/github/pr/RevenueCat/purchases-android/3660?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/RevenueCat/purchases-android/3660?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#3656** <a href="https://app.graphite.com/github/pr/RevenueCat/purchases-android/3656?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by AlvaroBrey** [comment]
@cursor review

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt [hunk 1]
```diff
@@ -1008,6 +1012,9 @@ internal class PaywallViewModelImpl(
         _workflowState.value = null
         if (isNewWorkflowImpression) {
             workflowTraceId = UUID.randomUUID().toString()
+            // Fresh presentation: start the shared store empty; each step registers its declarations as it builds.
```
[the remark this answers is not in the document]
Comment (AlvaroBrey) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt:1129]: Fixed in a93263143. The standalone path now reuses one `PaywallStateStore` (a ViewModel-owned `standaloneStateStore`, lazily created in `updatePaywallState`) across full state refreshes like color changes, instead of minting a fresh store each time via `toComponentsPaywallState`. It is tied to the ViewModel lifetime, so state still resets when the paywall is presented again, matching the workflow-shared store and iOS (`@StateObject`). Note this was latent in this PR since nothing reads or writes the store until the component wiring lands (Phase 1), but fixing the lifecycle now keeps it correct.

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt [hunk 2]
```diff
@@ -1130,6 +1145,7 @@ internal class PaywallViewModelImpl(
             colorScheme = _colorScheme.value,
             storefrontCountryCode = purchases.storefrontCountryCode,
             mode = options.mode,
+            stateStore = currentWorkflowStateStore,
```
[the remark this answers is not in the document]
Comment (AlvaroBrey) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt]: false positive

**Comment by AlvaroBrey** [comment]
@cursor review

**Comment by AlvaroBrey** [comment]
bugbot run

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallStateStore.kt [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import kotlinx.serialization.json.longOrNull
+import java.util.concurrent.ConcurrentHashMap
+
+/** In-memory state store for state-driven paywalls, scoped to one presentation session. */
+@OptIn(InternalRevenueCatAPI::class)
+@Stable
+internal class PaywallStateStore(declarations: Map<String, StateDeclaration>) {
+
+    private val declaredTypes = ConcurrentHashMap<String, StateDeclaration.ValueType>()
+    private val declaredDefaults = mutableStateMapOf<String, JsonPrimitive>()
```
Comment (MonikaMateska) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallStateStore.kt]: This is a nit: the declaredDefaults never mutates after registration, and subscribers callers to an observable map for no reason. I think Map should be sufficient here.

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt [hunk 4]
```diff
@@ -210,6 +210,13 @@ internal class PaywallViewModelImpl(
     private var currentWorkflowPresentedOfferingContext: PresentedOfferingContext? = null
     private var currentWorkflowStepTracksPaywallEvents = true
     private val workflowStepStateCache = mutableMapOf<String, PaywallState.Loaded.Components>()
+
+    // Shared across all screens of a workflow presentation so state-driven values survive screen navigation.
+    private var currentWorkflowStateStore: PaywallStateStore? = null
+
+    // Reused across full state refreshes (e.g. color changes) of a standalone paywall so its state-driven values
+    // are not lost. Tied to this ViewModel's lifetime, so it resets when the paywall is presented again.
+    private var standaloneStateStore: PaywallStateStore? = null
```
Comment (MonikaMateska) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt:220]: Can we add few unit tests to cover these behaviors from the view model:
- standaloneStateStore reuse across a color-scheme refresh
- currentWorkflowStateStore shared across screens
- prewarm swap-abandon (if (stateStore !== currentWorkflowStateStore) return@launch)

This is definitely non-blocking, but something nice to have

**Review by MonikaMateska — APPROVED** [review]
Really nice work! Just two small things that are non-blocking.

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallStateStore.kt (hunk 3 above)
Comment (AlvaroBrey) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallStateStore.kt]: Fixed in https://github.com/RevenueCat/purchases-android/pull/3660/changes/dd6eef21b82c36a9c2c14e8e83cb66607eac587a

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt (hunk 4 above)
Comment (AlvaroBrey) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt:220]: Added in https://github.com/RevenueCat/purchases-android/pull/3660/changes/dd6eef21b82c36a9c2c14e8e83cb66607eac587a

**Comment by AlvaroBrey** [comment]
### Merge activity

* **Jul 21, 9:48 AM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/RevenueCat/purchases-android/3660).
* **Jul 21, 9:49 AM UTC**: [Graphite](https://app.graphite.com/github/pr/RevenueCat/purchases-android/3660) rebased this pull request as part of a merge.
* **Jul 21, 10:04 AM UTC**: [Graphite](https://app.graphite.com/github/pr/RevenueCat/purchases-android/3660) couldn't merge this PR because it failed for an unknown reason (This repository has GitHub's merge queue enabled. Please configure the GitHub merge queue integration in Graphite settings.).
